source/Ship.cpp

```diff
@@ -3498,20 +3508,58 @@ const vector<weak_ptr<Ship>> &Ship::GetEscorts() const
 // cues and try to stay with it when it lands or goes into hyperspace.
 void Ship::AddEscort(Ship &ship)
 {
+	RegisterEscort(ship);
 	escorts.push_back(ship.shared_from_this());
 }
 
 
 
 void Ship::RemoveEscort(const Ship &ship)
 {
+	// Reset this value, we will re-scan the list of escorts to set the
+	// escorts velocity based on remaining active escorts.
+	escortsVelocity = -1.;
+
 	auto it = escorts.begin();
-	for( ; it != escorts.end(); ++it)
-		if(it->lock().get() == &ship)
+	while(it != escorts.end())
+	{
+		auto escort = it->lock().get();
+		if(escort)
 		{
-			escorts.erase(it);
-			return;
+			if(escort == &ship)
+			{
+				it = escorts.erase(it);
+				continue;
+			}
+			else
+				// Re-scan existing escorts to cache data like the
+				// escorts velocity.
+				RegisterEscort(ship);
 		}
+		++it;
+	}
+}
+
+
+
+// Store relevant cached data for the given escort.
+void Ship::RegisterEscort(const Ship &ship)
+{
+	// Cache the maximum speed that the parent should stay below to keep
+	// all escorts together.
+	// We don't cache the speeds of escorts that have 0 velocity to avoid
+	// having the parent get stuck when one of the escorts doesn't have
+	// regular thrust.
+	// We also don't cache the speeds of carried ships, since they are often
+	// docked and the carrier waits for them during docking already.
+	if(!ship.CanBeCarried() && ship.MaxVelocity() > 0.)
```

Comment:
Sorry, I wasn't really clear with the previous suggestion -- since MaxVelocity is a computed quantity, can we only compute it once, after we know it's not a carried ship?